File: nvpl_blas/c/cgbmv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_cgbmv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t M = 3;
    nvpl_int_t N = 4;
    nvpl_int_t KL = 1;
    nvpl_int_t KU = 2;
    nvpl_scomplex_t alpha = {1.0f, 2.0f};
    nvpl_scomplex_t beta = {1.0f, 2.0f};
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE transA = CblasNoTrans;
    nvpl_scomplex_t * A = NULL;
    nvpl_scomplex_t * X;
    nvpl_scomplex_t * Y;
    nvpl_int_t incX = 1;
    nvpl_int_t incY = 1;

    nvpl_int_t nx = (transA == CblasNoTrans) ? N : M;
    nvpl_int_t ny = (transA == CblasNoTrans) ? M : N;
    
    nvpl_int_t lda = KL + KU + 1;
    nvpl_int_t size_k = (order == CblasColMajor) ? N : M;
    nvpl_int_t len_x = 1 + (nx - 1) * labs(incX);
    nvpl_int_t len_y = 1 + (ny - 1) * labs(incY);

    printf("\nExample: cblas_cgbmv for banded matrix-vector multiplication\n\n");
    printf("#### args: m=%" PRId64 ", n=%" PRId64 ", kl=%" PRId64 ", ku=%" PRId64 ", lda=%" PRId64 ", "
            "incx=%" PRId64 ", incy=%" PRId64 ", transA=%c, order=%c\n", (int64_t)M, (int64_t)N, (int64_t)KL,
           (int64_t)KU, (int64_t)lda, (int64_t)incX, (int64_t)incY, transpose_to_char(transA),
           order_to_char(order));
    printf("           alpha=(%g, %g), beta=(%g, %g)\n", alpha.real, alpha.imag, beta.real, beta.imag);

    // allocate memory
    A = (nvpl_scomplex_t *)malloc(lda * size_k * sizeof(nvpl_scomplex_t));
    X = (nvpl_scomplex_t *)malloc(len_x * sizeof(nvpl_scomplex_t));
    Y = (nvpl_scomplex_t *)malloc(len_y * sizeof(nvpl_scomplex_t));

    // fill data
    fill_cbandmatrix(A, KL, KU, M, N, lda, order);
    fill_cvector(X, nx, incX);
    fill_cvector(Y, ny, incY);

    // print input data
    print_cbandmatrix(A, KL, KU, M, N, lda, "A", order);
    print_cvector(X, nx, incX, "X");
    print_cvector(Y, ny, incY, "Y");

    // call cblas_cgbmv
    cblas_cgbmv(order, transA, M, N, KL, KU, (void *)&alpha, A, lda, X, incX, (void *)&beta, Y, incY);

    // print result
    printf("\nAfter cgbmv operation:\n");
    print_cvector(Y, ny, incY, "Y");

    // release memory
    free(A);
    free(X);
    free(Y);
    return 0;
}

File: nvpl_blas/f77/cgbmv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     CGBMV
!
!******************************************************************************
      program   CGBMV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      character*1     transa
      parameter       (transa='N')
      integer         m, n, lda
      integer         incx, incy, i, nx, ny, xymax
      parameter       (m=3, n=4, incx=1, incy=1, xymax=4)
      integer         kl, ku
      parameter       (kl=1, ku=2, lda=kl+ku+1)
      integer         asize
      parameter       (asize=m*n)
      complex         alpha, beta
      parameter       (alpha=(1.0,2.0), beta=(1.0,2.0))
      complex         a(lda,n)
      integer         xsize, ysize
      parameter       (xsize=1+(xymax-1)*abs(incx),
     &                 ysize=1+(xymax-1)*abs(incy))
      complex         x(xsize), y(ysize)

*       External Subroutines
      external        print_cbandmatrix, fill_cbandmatrix,
     &                print_cvector, fill_cvector
*
*      Executable Statements
*
      print*
      print 99, 'Example: CGBMV for banded matrix-vector ',
     &          'multiplication'
*
*       Print input data
      print*
      print 100, m, n, kl, ku, lda, incx, incy, transa
      print 101, alpha, beta
      if (transa .eq. 'N') then
          nx=n
          ny=m
      else
          nx=m
          ny=n
      endif
      call fill_cbandmatrix(a,kl,ku,m,n,lda)
      call fill_cvector(x,nx,incx)
      call fill_cvector(y,ny,incy)
      call print_cbandmatrix(a,kl,ku,m,n,lda,'A')
      call print_cvector(x,nx,incx,'X')
      call print_cvector(y,ny,incy,'Y')
*
*      Call CGBMV subroutine
      call CGBMV(transa,m,n,kl,ku,alpha,a,lda,x,incx,beta,y,incy)
*
      print*
      print 102, 'After cgbmv operation:'
      call print_cvector(y,ny,incy,'Y')

      stop
 99   format(A,A)
 100  format('#### args: m=',i1,', n=',i1,', kl=',i1,', ku=',i1,
     &       ', lda=',i1,', incx=',i1,', incy=',i1,', transA=',a1)
 101  format(11x,'alpha=(',f3.1,', ',f3.1,'),',
     &       ' beta=(',f3.1,', ',f3.1,')')
 102  format(A)
      end
